File: Core/Kernel/TrustCacheManager.m
#import "TrustCacheManager.h"
#import "darksword.h"
#import "choma_helpers.h"
#import "choma_trustcache.h"
#import "ChoMAWrapper.h"
#import "Logger.h"

@implementation TrustCacheManager

+ (BOOL)injectTrustCacheWithCdhash:(NSData *)cdhash error:(NSError **)error {
    if (!ds_is_ready()) {
        LOG_ERROR("'KRW not ready");
        if (error) *error = [NSError errorWithDomain:@"TC" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"KRW not ready"}];
        return NO;
    }
    if (cdhash.length != 20) {
        LOG_ERROR("'Invalid CDHash length: %lu", (unsigned long)cdhash.length);
        if (error) *error = [NSError errorWithDomain:@"TC" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"Invalid CDHash length"}];
        return NO;
    }

    uint64_t kernelSlide = ds_get_kernel_slide();
    LOG_DEBUG("Kernel slide = 0x%llx", kernelSlide);

    uint64_t tcUnslid = choma_find_trust_cache_by_data_scan();
    if (tcUnslid == 0) {
        LOG_ERROR("Trust cache not found via data scan");
        if (error) *error = [NSError errorWithDomain:@"TC" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"Trust cache not found"}];
        return NO;
    }
    uint64_t tcAddr = tcUnslid + kernelSlide;
    LOG_DEBUG("'TC addr (unslid=0x%llx, slid=0x%llx)", tcUnslid, tcAddr);

    uint32_t version = ds_kread32(tcAddr);
    if (version != 1 && version != 2) {
        LOG_ERROR("'Unknown TC version: %u", version);
        if (error) *error = [NSError errorWithDomain:@"TC" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"Unknown TC version"}];
        return NO;
    }
    uint32_t count = ds_kread32(tcAddr + 20);
    LOG_DEBUG("'TC version=%u, current entries=%u", version, count);

    uint64_t amfiDataSize = 0;
    uint64_t amfiDataStart = choma_get_amfi_data_range(&amfiDataSize);
    if (amfiDataStart == 0 || amfiDataSize == 0) {
        LOG_ERROR("'AMFI data range not found");
        if (error) *error = [NSError errorWithDomain:@"TC" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"AMFI data range not found"}];
        return NO;
    }
    uint64_t dataEnd = amfiDataStart + kernelSlide + amfiDataSize;
    LOG_DEBUG("'AMFI data: start=0x%llx size=%llu end=0x%llx", amfiDataStart, amfiDataSize, dataEnd);

    uint64_t cdhashArray = tcAddr + 24;
    uint64_t newEntryOff = cdhashArray + (uint64_t)count * 20;
    LOG_DEBUG("'New entry would be at 0x%llx (data ends at 0x%llx)", newEntryOff, dataEnd);

    if (newEntryOff + 20 > dataEnd) {
        LOG_ERROR("'No slack space: entry 0x%llx + 20 > data end 0x%llx", newEntryOff, dataEnd);
        if (error) *error = [NSError errorWithDomain:@"TC" code:-1 userInfo:@{NSLocalizedDescriptionKey: @"No slack space in TC data section"}];
        return NO;
    }

    // Dedup
    for (uint32_t i = 0; i < count; i++) {
        uint8_t buf[20];
        ds_kreadbuf(cdhashArray + (uint64_t)i * 20, buf, 20);
        NSData *existing = [NSData dataWithBytes:buf length:20];
        if ([existing isEqualToData:cdhash]) {
            LOG_INFO("CDHash already in trust cache (entry %u) — skipping", i);
            return YES;
        }
    }

    LOG_DEBUG("'Writing CDHash at 0x%llx and incrementing count...", newEntryOff);
    ds_kwritebuf(newEntryOff, cdhash.bytes, 20);
    ds_kwrite32(tcAddr + 20, count + 1);
    LOG_INFO("Trust cache injected — new entry count: %u", count + 1);

    return YES;
}

@end

File: Core/Kernel/TrustCacheManager.swift
import Foundation

class TrustCacheManager {
    // Lấy CDHash từ file binary của IPA (dùng ChOma)
    static func getCDHash(from machOPath: URL) -> Data? {
        return ChoMAWrapper.extractCDHash(machOPath.path)
    }

    // Inject CDHash vào Kernel Memory
    static func injectTrustCache(cdhash: Data) throws {
        guard ds_is_ready() else { throw KernelError.exploitFailed("KRW not ready") }
        guard cdhash.count == 20 else { throw KernelError.cdHashExtractionFailed }

        // Tìm trust cache trong kernel memory qua XPF scan
        // (thay thế off_g_system_trust_cache — không có trong offsets.h)
        let kernelSlide = ds_get_kernel_slide()
        let tcUnslid = choma_find_trust_cache_by_data_scan()
        guard tcUnslid > 0 else { throw KernelError.exploitFailed("Trust cache not found") }

        let tcAddr = tcUnslid + kernelSlide

        // Đọc header: version(4) + uuid(16) + count(4)
        let version = ds_kread32(tcAddr)
        guard version == 1 || version == 2 else { throw KernelError.exploitFailed("Unknown TC version") }
        let count = ds_kread32(tcAddr + 20)

        // Tìm vùng nhớ an toàn trong data section (đã có sẵn slack space)
        var amfiDataSize: UInt64 = 0
        let amfiDataStart = choma_get_amfi_data_range(&amfiDataSize)
        guard amfiDataStart > 0, amfiDataSize > 0 else { throw KernelError.exploitFailed("AMFI data not found") }
        let dataEnd = amfiDataStart + kernelSlide + amfiDataSize

        // Ghi CDHash vào cuối mảng cdhash trong trust cache
        let cdhashArray = tcAddr + 24
        let newEntryOff = cdhashArray + UInt64(count) * 20

        // Check bounds: còn chỗ không?
        guard newEntryOff + 20 <= dataEnd else { throw KernelError.exploitFailed("No slack space in TC") }

        // Dedup: kiểm tra CDHash đã tồn tại chưa
        for i in 0..<count {
            var existing = Data(count: 20)
            _ = existing.withUnsafeMutableBytes { buf in
                ds_kreadbuf(cdhashArray + UInt64(i) * 20, buf.baseAddress, 20)
            }
            if existing == cdhash { return } // đã tồn tại
        }

        // Ghi CDHash mới
        cdhash.withUnsafeBytes { buf in
            ds_kwritebuf(newEntryOff, buf.baseAddress, 20)
        }

        // Tăng count
        ds_kwrite32(tcAddr + 20, count + 1)
    }
}
